### DISCUSSION TITLE
Colaboraciones con datos insuficientes

### DISCUSSION THREAD

**Discussion by Traspa1112** [dds-jv/foro#424] [TPA - 2024] [answered]
Buenas tardes. Mientras intentábamos satisfacer el requerimiento "Carga masiva de colaboraciones" nos encontramos con un problema al momento de intentar generar las colaboraciones correspondientes, ya que para todas nos falta información. Incluso si pudieramos generar dichas colaboraciones *incompletas*, existiría un problema aún mayor al momento de calcular el puntaje de los colaboradoradores cargados masivamente, ya que en la fórmula dada requiere varios datos faltantes. Por ejemplo sería imposible determinar las semanas frescas de una vianda o los meses que estuvo activa una tarjeta repartida.
Teniendo en cuenta que en reiteradas ocasiones se nos indica consultar al analista, es que decidimos generar esta discusión con la intención de solventar dichas dificultades.
Quedamos a la espera de su respuesta. Saludamos atentamente, Tomás y Agustín.

**Comment by Traspa1112** [comment]
Buenas noches. Releyendo los requerimientos detectamos un nuevo incoveniente relacionado con los anteriores. Si el .csv contuviese una colaboración de tipo "ENTREGA_TARJETAS" de un colaborador que no existe previamente o de quien se desconoce su domicilio, el sistema entraría en un estado inválido ya que no se cuenta con el domicilio del mismo entre los datos de la carga masiva. Dicha información, según los requerimientos, se considera requerida para los colaboradores que decidan repartir tarjetas.
Nuevamente, en su rol de analista, les pedimos nos ayuden a esclarecer este conflicto en los requerimientos.
Quedamos a la espera de su respuesta. Saludamos atentamente, Agustín y Tomás.

**Reply by flbulgarelli** [reply]
>  Si el .csv contuviese una colaboración de tipo "ENTREGA_TARJETAS" de un colaborador que no existe previamente o de quien se desconoce su domicilio, el sistema entraría en un estado inválido ya que no se cuenta con el domicilio del mismo entre los datos de la carga masiva

Es posible que el CSV cuente con datos incorrectos, como por ejemplo en el caso que señalás (en que refiere a una persona que no está habilitada para hacer ese tipo de colaboración). En todas estas situaciones deberías validar la situación y o bien (y a criterio del grupo y a discutir con su ayudante en la entrega):

 1. Fallar ante el primer error sin hacer **ninguna** modificación al sistema
 2. Ignorar las filas que representen información inválida, procesar las restantes, e informar al finalizar (retornando un objeto apropiado) cuales fueron las filas no procesadas y por qué.

**Comment by flbulgarelli** [comment]
:wave: ¡Hola!
 
> Buenas tardes. Mientras intentábamos satisfacer el requerimiento "Carga masiva de colaboraciones" nos encontramos con un problema al momento de intentar generar las colaboraciones correspondientes, ya que para todas nos falta información. Incluso si pudieramos generar dichas colaboraciones _incompletas_, existiría un problema aún mayor al momento de calcular el puntaje de los colaboradoradores cargados masivamente, ya que en la fórmula dada requiere varios datos faltantes. 

En general, la idea es que se creen las colaboraciones únicamente con la información del CSV, utilizando valores por defecto razonables en los casos en que la misma no sea parte de la especificación del CSV. ¿Cuáles son los valores que considerás que faltan y no podés calcular o proveer por ningún otro medio? Contame eso y te respondo por los casos puntuales.

**Reply by agmblanco** [reply]
Buenos días. Son varios. Por un lado, como dijimos en el primer mensaje, las colaboraciones de donación de vianda cargadas de forma masiva no cuentan con información de la vianda, por lo que, entre otras cosas, no tendríamos forma de determinar la cantidad de `semanas fresca` de cada una de esas; dato que resulta requerido en la fórmula para calcular el puntaje del colaborador. 
De la misma manera, sería imposible determinar las `semanas activa` de una tarjeta repartida en una colaboración de tipo "Registro de persona vulnerable" que se reciba como dato en el CSV de carga masiva. Dato que también resulta requerido para satisfacer la fórmula provista para el cálculo de puntajes.

>Es posible que el CSV cuente con datos incorrectos, como por ejemplo en el caso que señalás (en que refiere a una persona que no está habilitada para hacer ese tipo de colaboración).

Con respecto a esto, surge un problema adicional ya que en los requerimientos se pide que de no existir el colaborador en el sistema se cree, pero en el caso de reparto de tarjetas, como dijimos llevaría inevitablemente a un estado incorrecto. Teniendo en cuenta esto, surgen, a nuestro criterio, dos posibles alternativas:
a) Cargamos las colaboraciones como dato histórico pero configuramos al colaborador creado de forma tal que no tenga permitido realizar este tipo de colaboraciones (Reparto de tarjetas/Registro persona vulnerable) en el futuro.
b) En caso de recibir una colaboración de este tipo, _Reparto de tarjetas/Registro persona vulnerable_, de un colaborador que no se encuentra cargado en el sistema, devolver alguno de los estados de error sugeridos en la primer respuesta (Deshacer toda la transacción ó ignorar este caso particular y devolver un listado de los datos no insertados)

**Comment by D-M-Gonzalez** [comment]
Buenas, aprovecho para sumarme a esta discusión, ya que tuvimos una duda similar, consultamos y todavía no obtuvimos respuesta de la ayudante.
Nosotros asumimos que las colaboraciones quedarían vacías, y que en el caso del puntaje las ignoramos, por lo menos por ahora hasta saber que sucede.

Esta es la fórmula para el puntaje:

[PESOS_DONADOS] * 0,5 +
[VIANDAS_DISTRIBUIDAS] * 1 +
[VIANDAS_DONADAS] * [∑ SEMANAS_FRESCA] *]1,5 +
[TARJETAS_REPARTIDAS] * [∑ MESES_ACTIVAS] * [∑ USOS]  * 2 +
[CANTIDAD_HELADERAS_ACTIVAS] * [∑ MESES_ACTIVAS] *  [∑ USOS] * 5

Pero en el CSV nos falta para el caso de:
- Donación Dinero: No tenemos la cantidad de pesos donados.
- Distribución de viandas: No tenemos la cantidad de viandas distribuidas.
- Viandas donadas: No tenemos el vencimiento de la vianda.
- Distribución de tarjetas: No tenemos datos de la persona en situación de riesgo para darle el alta y ver cuantas veces uso la tarjeta.
- Administración de heladeras: No tenemos datos de la heladera para darle el alta, por lo que nunca van a estar activas.

El CSV solo contiene datos del colaborador y no de las colaboraciones.

**Reply by flbulgarelli** [reply]
> El CSV solo contiene datos del colaborador y no de las colaboraciones.

Es (mayormente) correcto: tiene información del colaborador, y sólo algunos datos de las colaboraciones.  Entiendo que todas estas preguntas quedan respondidas con mis respuestas posteriores.

**Comment by flbulgarelli** [comment, marked as answer]
¡Hola!

Es importante tener en cuenta el contexto que plantea para la funcionalidad de CSV: 

> Dado que este Sistema no es un programa nuevo de la ONG, se solicitó poder hacer la migración de los donantes que ya existen pero de manera masiva. 

Es decir, esta carga de información es sobre donaciones pasadas, anteriores a la creación del sistema software que se está desarrollando, por lo que podemos asumir que: 

1. Estamos hablando siempre de colaboraciones en el pasado, por lo que no utilizaremos la información para gestionar datos (ni viandas) "frescos". En otras palabras, las colaboraciones que se cargarán serán mucho más distantes en el tiempo que lo que dura el ciclo de vida de una vianda. 
2. La información que se ingresa es a fines históricos, y lo que se busca es poder tener la mayor cantidad de información pasada, a fines de registro y auditoría, pero nuevamente, no se pretende contar con esa información para realizar acciones o tomar decisiones operativas y de corto plazo. Además, cuando nos encontramos en este tipo de situaciones, se pretenderá hacer el mayor esfuerzo posible para reconstituir la historia pasada, pero se da por sentado que nunca se contará al final de ese con información tan precisa como la que se carga con el sistema nuevo: la información será incompleta y/o aproximada.

**Reply by flbulgarelli** [reply]
Dicho ésto, vamos a las dudas particulares: 

* (...) _las colaboraciones de donación de vianda cargadas de forma masiva no cuentan con información de la vianda, por lo que, entre otras cosas, no tendríamos forma de determinar la cantidad de semanas fresca de cada una de esas_ (...): correcto, no se cuenta con dicha información, porque anteriormente en el sistema no se cargaba. En estos casos, simplemente, no se sabe cuál fue la vianda donada. 
* (...) _dato que resulta requerido en la fórmula para calcular el puntaje del colaborador_: el cálculo del puntaje para la información histórica ingresada por CSV será por definición impreciso, porque no se cuenta con estos datos. Podés asumir que estas colaboraciones no computan para el cálculo, o bien que computan un valor fijo (por ejemplo, asumir, arbitrariamente, que estas viandas "fantasma" duraron siempre una semana frescas). Incluso podrías ir más allá y asumir que este cálculo sólo estará disponible para colaboraciones hechas después de una cierta fecha X, y luego validar que las colaboraciones ingresadas por CSV deban ser siempre anteriores a dicha fecha X. Todas estas opciones son, en principio, igualmente válidas y a conversar con tu ayudante. 
*  _De la misma manera, sería imposible determinar las semanas activa de una tarjeta repartida_ (...): mismas consideraciones que en el punto anterior. 
* _como dijimos llevaría inevitablemente a un estado incorrecto_: ese estado, para colaboraciones ingresadas por este medio, no es incorrecto, sino que debés contemplarlo (aunque, nuevamente, por el tipo de casos de uso, es posible que no tengan un impacto significativo en las operaciones del sistema).

**Reply by flbulgarelli** [reply]
Por último, sobre ésto...

> a) Cargamos las colaboraciones como dato histórico pero configuramos al colaborador creado de forma tal que no tenga permitido realizar este tipo de colaboraciones (Reparto de tarjetas/Registro persona vulnerable) en el futuro.
b) En caso de recibir una colaboración de este tipo, Reparto de tarjetas/Registro persona vulnerable, de un colaborador que no se encuentra cargado en el sistema, devolver alguno de los estados de error sugeridos en la primer respuesta (Deshacer toda la transacción ó ignorar este caso particular y devolver un listado de los datos no insertados)

... definitivamente, por lo que te comentaba antes, estas situaciones de ausencia "estructural" de ciertos datos de la colaboración que estamos conversando no representan un caso de error, por lo que no aplica para las mismas la alternativa _b_ que describís (aunque sí puede aplicar para otras situaciones, como por ejemplo, un campo DNI vacío).

**Reply by D-M-Gonzalez** [reply]
Muchas gracias por la respuesta! eso termina de aclarar nuestras dudas. Igual no marco como respuesta porque no es mi publicación.

**Reply by agmblanco** [reply]
Buenas noches Franco, gracias por las respuestas.

> _Podés asumir que estas colaboraciones no computan para el cálculo, o bien que computan un valor fijo [...]_

Si, esa fue una de las consultas planteadas en el e-mail, pero no estábamos seguros sobre su validez ya que en el pie de la sexta página del enunciado decía _"En caso de donación de viandas no poseemos datos sobre las mismas, sin embargo se debe registrar al colaborador (si es que aún no estaba registrado) y computar los puntos de todas formas"_. Teniendo en cuenta que es una alternativa viable repensaremos esa parte. 

> _Todas estas opciones son, en principio, igualmente válidas y a conversar con tu ayudante._

Si, en este caso no sabíamos donde correspondía hacer la consulta por lo que también le enviamos un e-mail con otras dudas que mayormente quedan resueltas con las respuestas de este foro. Tanto Tomás como yo pertenecemos a un grupo bajo su tutela. 

Para concluir, lo repensaremos teniendo en cuenta las respuestas recibidas y en caso de seguir habiendo dudas mañana lo charlaremos personalmente.

Saludos cordiales, Agustín M. Blanco.
